File: rb/test/event_entity_test.rb
# Event entity test

require "minitest/autorun"
require "json"
require_relative "../Openfda_sdk"
require_relative "runner"

class EventEntityTest < Minitest::Test
  def test_create_instance
    testsdk = OpenfdaSDK.test(nil, nil)
    ent = testsdk.Event(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = event_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "event." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set OPENFDA_TEST_EVENT_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    event_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.event")))
    event_ref01_data = nil
    if event_ref01_data_raw.length > 0
      event_ref01_data = Helpers.to_map(event_ref01_data_raw[0][1])
    end

    # LIST
    event_ref01_ent = client.Event(nil)
    event_ref01_match = {}

    event_ref01_list_result, err = event_ref01_ent.list(event_ref01_match, nil)
    assert_nil err
    assert event_ref01_list_result.is_a?(Array)

  end
end

def event_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "event", "EventTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = OpenfdaSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["event01", "event02", "event03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["OPENFDA_TEST_EVENT_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "OPENFDA_TEST_EVENT_ENTID" => idmap,
    "OPENFDA_TEST_LIVE" => "FALSE",
    "OPENFDA_TEST_EXPLAIN" => "FALSE",
    "OPENFDA_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["OPENFDA_TEST_EVENT_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["OPENFDA_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["OPENFDA_APIKEY"],
      },
      extra || {},
    ])
    client = OpenfdaSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["OPENFDA_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["OPENFDA_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: php/test/EventEntityTest.php
<?php
declare(strict_types=1);

// Event entity test

require_once __DIR__ . '/../openfda_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class EventEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = OpenfdaSDK::test(null, null);
        $ent = $testsdk->Event(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "event" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = OpenfdaSDK::test($seed, null);
        $seen = iterator_to_array($base->Event(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = OpenfdaConfig::make_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = OpenfdaSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->Event(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = event_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "event." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set OPENFDA_TEST_EVENT_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $event_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.event")));
        $event_ref01_data = null;
        if (count($event_ref01_data_raw) > 0) {
            $event_ref01_data = Helpers::to_map($event_ref01_data_raw[0][1]);
        }

        // LIST
        $event_ref01_ent = $client->Event(null);
        $event_ref01_match = [];

        $event_ref01_list_result = $event_ref01_ent->list($event_ref01_match, null);
        $this->assertIsArray($event_ref01_list_result);

    }
}

function event_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/event/EventTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = OpenfdaSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["event01", "event02", "event03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("OPENFDA_TEST_EVENT_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "OPENFDA_TEST_EVENT_ENTID" => $idmap,
        "OPENFDA_TEST_LIVE" => "FALSE",
        "OPENFDA_TEST_EXPLAIN" => "FALSE",
        "OPENFDA_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["OPENFDA_TEST_EVENT_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["OPENFDA_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["OPENFDA_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new OpenfdaSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["OPENFDA_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["OPENFDA_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
